File: 프로그래머스/lv2/87946. 피로도/피로도.py
def dfs(dungeons, visited, life):
    N = len(dungeons)
    global answer
    
    for i in range(N):
        if visited[i] == 1:
            continue
        if dungeons[i][0] > life:
            continue
        else:
            visited[i] = 1
            dfs(dungeons, visited, life-dungeons[i][1])
            visited[i] = 0
            
    answer = max(answer, sum(visited))
    
def solution(k, dungeons):
    N = len(dungeons)
    visited=[0]*N
    global answer
    answer = 0
    
    for i in range(N):
        dfs(dungeons, visited, k)
    
    return answer
    
        
        

File: 프로그래머스/2/87946. 피로도/피로도.java
import java.util.*;

class Solution {
    
    static boolean[] visited;
    static int max = 0;
    
    public int solution(int k, int[][] dungeons) {
        visited = new boolean[dungeons.length];
        dfs(k, dungeons, 0);
        return max;
    }
    
    public void dfs(int current, int[][] dungeons, int count) {        
        for (int i = 0; i < dungeons.length; i++) {
            if (visited[i]) {
                continue;
            }
            
            if (dungeons[i][0] <= current) {
                visited[i] = true;
                dfs(current - dungeons[i][1], dungeons, count + 1);
                visited[i] = false;
            } 
        }
        
        max = Math.max(max, count);
    }
}
